### ISSUE TITLE
[BUG] Exercise in Chapter 10: Test Error at 9.8% instead of ~98%

### ISSUE THREAD

**Issue by Sam-OT** [ageron/handson-ml2#483]
**Describe the bug**

I was going through the coding exercise in Chapter 10. I got to almost the very end, cell 127, and I get a completely different accuracy. I get about 9.8%, rather than 98%, accuracy! Confusingly, the _loss_ is very similar: 0.082 instead of 0.080.

The previous cell runs very similarly to the output in the GitHub page, getting to a bit over 98% accuracy on the validation set. It's not quite the same as yours, but I guess that this is due to a different version number or something. I have saved the notebook, which I can provide if required. I'm not exactly sure how to do so here, though.

Here is the output from my run.

![image](https://user-images.githubusercontent.com/81111289/137115416-6a8e2479-567d-4088-809c-24cbcfefc4a6.png)

Here is what your GitHub page shows.

![image](https://user-images.githubusercontent.com/81111289/137115919-3ab76438-4ce9-4ad2-8533-b4930b35786f.png)

I installed with Anaconda and got v2.3.0, not v2.4.1, like you have. I created this with `conda -n TensorFlowEnv tensorflow numpy pandas matplotlib seaborn spyder jupyter`. This gave my python 3.8; I'm not sure which you used.

Anyway, the accuracy on the test set is roughly 10%, rather than roughly 100%. I downloaded your exact Jupyter Notebook and re-ran it, to make sure that it wasn't a typo on my behalf. The same (unexpected) behaviour occured.

I wondered if it was maybe a pre-processing step which hadn't been done for the test data, but had for the training and validation. But this doesn't appear to be the case. See cell 100.

Not _everything_ else was _exactly_ as in your notebook output. I guess this is, again, due to the different version. But they're fundamentally basically the same. Eg, one loss is 0.2255 instead of 0.2252, or something like that.

**To Reproduce**

describe above

**Expected behavior**

described above

**Screenshots**

given above

**Versions (please complete the following information):**
 - OS: Windows 10
 - Python: 3.8
 - TensorFlow: 2.3.0
 - Scikit-Learn: 0.24.2

**Comment by Sam-OT** [comment]
I've done a bit more experimentation. Interestingly, if I drop the whole "save and then load best model" part, then it works as expected. I got 98.5% accuracy on the test set. Here's a minimal working example, taking the appropriate parts from your notebook.

```
import tensorflow as tf
from tensorflow.keras import layers

tf.random.set_seed(42)

(X_train_full, y_train_full), (X_test, y_test) = tf.keras.datasets.mnist.load_data()
X_valid, X_train = X_train_full[:5000] / 255., X_train_full[5000:] / 255.
y_valid, y_train = y_train_full[:5000],        y_train_full[5000:]
X_test = X_test / 255.

model = tf.keras.models.Sequential([
    layers.Flatten(input_shape = [28, 28]),
    layers.Dense(300, activation = "relu"),
    layers.Dense(100, activation = "relu"),
    layers.Dense(10,  activation = "softmax")
])
model.compile(
    loss = "sparse_categorical_crossentropy",
    optimizer = tf.keras.optimizers.SGD(learning_rate = 0.3),
        #   this is the learning rate that you chose after the tuning
    metrics = ["accuracy"]
)

early_stopping_cb = tf.keras.callbacks.EarlyStopping(patience = 20)
history = model.fit(
    X_train, y_train, epochs = 100,
    validation_data = (X_valid, y_valid),
    callbacks = [early_stopping_cb]
)

model.evaluate(X_test, y_test)
# ~ output: [0.0931241512298584, 0.9848999977111816]
```

I don't know what is going wrong, but something must be going wrong when you save the model then load it. I haven't been able to determine what the problem is, though.

**Comment by ageron** [comment]
On a fresh Windows VM I just installed the [latest Miniconda](https://docs.conda.io/en/latest/miniconda.html) package, then I opened the Anaconda prompt and ran:

```bash
conda update conda
conda install git
git clone https://github.com/ageron/handson-ml2.zip
cd handson-ml2
conda env create -f environment.yml
conda activate tf2
python -m ipykernel install --user --name=python3
jupyter notebook
```

There was an issue with Box2D: it needed to be installed using conda instead of pip, so I fixed this. Otherwise everything went fine. Could you please pull the latest `environment.yml` and give this a try?

**Comment by ageron** [comment]
I'm glad you got a good environment up and running!

Yes, the `channels` section in this project's `environment.yml` file does force `conda-forge` to be looked at first, before the defaults. You can also create a `.condarc` file in your home directory, with the following content:

```
channel_priority: strict
channels:
  - conda-forge
  - defaults
```

This will ensure that when you `conda install` something, the conda-forge channel gets looked up first.

Now I checked the versions of TensorFlow available on conda-forge and defaults right now:

**defaults channel**

|Package|MacOS|Windows|Linux
|--------|--------|--------|--------
|TensorFlow| <strike>2.0.0</strike>|**2.6.0**|**2.6.0**
|Matplotlib| 3.4.2| 3.4.2| 3.4.2

**conda-forge channel**

|Package|MacOS|Windows|Linux
|--------|--------|--------|--------
|TensorFlow| **2.6.0**|<strike>1.14.0</strike>|**2.6.0**
|Matplotlib| 3.4.3|3.4.3|3.4.3

So apparently on Windows TensorFlow 2.6.0 must come from the defaults channel, which is no doubt why the channel is not displayed in your case. And since conda-forge is listed first in `environment.yml`, it pulled Matplotlib 3.4.3 from conda-forge.

Hope this helps!

**Comment by ageron** [comment]
I'm glad everything's finally working well! Thanks again for your detailed feedback, it's was very helpful. 👍
